### PR TITLE
feat(cart): add state

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Fixes: N/A

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

Changes the names of existing cart selectors.

## Other information

### CODE REVIEW

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+      const expectedState = {
+        ...initialState,
+        cart: {
+          ...initialState.cart,
+          items: [],
+          ...cart
+        },
+        loading: false
+      }
+      const cartResetSuccess = new DaffCartClearSuccess(cart);
```
Comment (lderrickable) [inline libs/cart/src/reducers/cart/cart.reducer.spec.ts]: should be cartClearSuccess

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+      }
+      const cartReset = new DaffCartClear();
+      const result = reducer(initialState, cartReset);
+
+      expect(result).toEqual(expectedState);
+    });
+  });
+
+  describe('when CartClearSuccessAction is triggered', () => {
+    it('should guarantee there are no items in the cart', () => {
```
Comment (lderrickable) [inline libs/cart/src/reducers/cart/cart.reducer.spec.ts]: The text here should probably be something like: "should set the cart payload on state". I don't see anything in this test about guaranteeing that there are no items, because that happens in the driver.

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+  });
+
+  describe('when CartClearAction is triggered', () => {
+    it('should indicate that the cart is loading', () => {
+      const expectedState = {
+        loading: true,
+        cart: null,
+        errors: []
+      }
+      const cartReset = new DaffCartClear();
```
Comment (lderrickable) [inline libs/cart/src/reducers/cart/cart.reducer.spec.ts]: should be `const cartClear`

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+      }
+
+      error = 'error';
+
+      const addToCartFailure = new DaffAddToCartFailure(error);
+
+      result = reducer(state, addToCartFailure);
+    });
+
+    it('should setloading to false', () => {
```
Comment (lderrickable) [inline libs/cart/src/reducers/cart/cart.reducer.spec.ts]: I like your other wording better: "should indicate that the cart is not loading"

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+      const addToCartActionSuccess: DaffAddToCartSuccess = new DaffAddToCartSuccess(cart);
+      state = {
+        ...initialState,
+        loading: true
+      }
+
+      result = reducer(state, addToCartActionSuccess);
+    });
+
+    it('should setcart from action.payload', () => {
```
Comment (lderrickable) [inline libs/cart/src/reducers/cart/cart.reducer.spec.ts]: A lot of typos like this in this file: `setcart`

File: libs/cart/src/effects/cart.effects.spec.ts [hunk 6]
```diff
@@ -35,60 +45,111 @@ describe('Daffodil | Cart | CartEffects', () => {
...
+        expected = cold('--b', { b: cartCreateFailureAction });
+      });
+
+      it('should dispatch a CartCreateFailure action', () => {
+        expect(effects.create$).toBeObservable(expected);
+      });
+    });
+  });
+
+  describe('when CartCreateSuccessAction is triggered', () => {
```
Comment (lderrickable) [inline libs/cart/src/effects/cart.effects.spec.ts]: I think this test could be better. Take a look at what I send you in slack. You should be able to test that the store service is actually called.

File: libs/cart/src/effects/cart.effects.spec.ts [hunk 7]
```diff
@@ -128,35 +187,33 @@ describe('Daffodil | Cart | CartEffects', () => {
       });
     });
   });
-  describe('when CartResetAction is triggered', () => {
-
+  describe('when CartClearAction is triggered', () => {
     let expected;
-    const resetAction = new DaffCartReset();
+    const resetAction = new DaffCartClear();
```
Comment (lderrickable) [inline libs/cart/src/effects/cart.effects.spec.ts]: `resetAction` should be clearAction

File: libs/cart/src/effects/cart.effects.spec.ts [hunk 8]
```diff
@@ -128,35 +187,33 @@ describe('Daffodil | Cart | CartEffects', () => {
...
 
     describe('and the clear call to driver is successful', () => {
-
       beforeEach(() => {
-        spyOn(daffDriver, 'clear').and.returnValue(of(mockCart));
-        const resetCartSuccessAction = new DaffCartResetSuccess(mockCart);
+        daffDriverSpy.clear.and.returnValue(of(mockCart));
+        const resetCartSuccessAction = new DaffCartClearSuccess(mockCart);
```
Comment (lderrickable) [inline libs/cart/src/effects/cart.effects.spec.ts]: should be clearCartSuccessAction

File: libs/cart/src/selectors/cart.selector.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+export const cartStateSelector: MemoizedSelector<object, State['cart']> = createSelector(
+  selectCartState,
+  (state: State) => state.cart
+)
+
+export const selectCartValue: MemoizedSelector<object, CartState['cart']> = createSelector(
+  cartStateSelector,
+  (state: CartState) => state.cart
+)
```
Comment (lderrickable) [inline libs/cart/src/selectors/cart.selector.ts]: There should also be selectors for every other major attribute of the cart model. So like

selectCartId
selectCartSubtotal
selectCartGrandTotal
selectCartCoupons
selctCartItems

and then more selectors for the rest of the cart model:
  billing_address: DaffCartAddress | null;
  shipping_address: DaffCartAddress | null;
  payment?: DaffCartPaymentMethod;
  totals: DaffCartTotal[];
  shipping_information: DaffCartShippingInformation | null;
  available_shipping_methods: DaffCartShippingRate[];
  available_payment_methods: DaffCartPaymentMethod[];

Comment (lderrickable) [reply libs/cart/src/selectors/cart.selector.ts]: Then these will need to be added to the cart facade too

File: libs/cart/src/effects/cart.effects.ts [hunk 10]
```diff
@@ -11,64 +11,91 @@ import {
...
-    private storage: DaffCartStorageService) { }
+    private storage: DaffCartStorageService
+  ) {}
+
+  @Effect()
+  create$ = this.actions$.pipe(
+    ofType(DaffCartActionTypes.CartCreateAction),
+    switchMap((action: DaffCartCreate) =>
+      this.driver.create().pipe(
+        map(resp => {
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/cart/src/effects/cart.effects.ts]: These kinds of changes are probably needed all throughout this PR. I'm going to try not to comment on them all, so feel free to change this wherever it happens

[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/effects/cart.effects.ts]: Okay I'll change all instances that I find

File: libs/cart/src/reducers/cart-shipping-information/cart-shipping-information.reducer.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    case DaffCartShippingInformationActionTypes.CartShippingInformationDeleteAction:
+      return { ...state, loading: true };
+
+    case DaffCartShippingInformationActionTypes.CartShippingInformationLoadSuccessAction:
+      return {
+        ...state,
+        cart: {
+          ...state.cart,
+          // TODO: remove workaround
+          shipping_information: {...action.payload, address_id: null}
```
Comment (lderrickable) [inline libs/cart/src/reducers/cart-shipping-information/cart-shipping-information.reducer.ts]: I suppose the comment is explaining the `address_id: null`. But why is this here?

File: libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+      return { ...state, loading: true };
+
+    case DaffCartShippingAddressActionTypes.CartShippingAddressLoadSuccessAction:
+      return {
+        ...state,
+        cart: {
+          ...state.cart,
+          shipping_address: action.payload
+        },
+        loading: false
```
Comment (lderrickable) [inline libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts]: Maybe success actions should reset the errors to []? What do you think?

File: libs/cart/src/effects/cart-payment-methods.effects.ts [hunk 13]
```diff
@@ -0,0 +1,39 @@
+import { Injectable, Inject } from '@angular/core';
+import { switchMap, map, catchError, mapTo, tap } from 'rxjs/operators';
```
Comment (lderrickable) [inline libs/cart/src/effects/cart-payment-methods.effects.ts]: unused imports

File: libs/cart/src/reducers/cart-shipping-information/cart-shipping-information.reducer.ts (hunk 11 above)
Comment (griest024) [reply libs/cart/src/reducers/cart-shipping-information/cart-shipping-information.reducer.ts]: Becuase the shipping information driver returns a `DaffCartShippingRate` not a `DaffCartShippingInformation`. I left a comment in slack suggesting that we change the driver to return the shipping info as it would eliminate workarounds like this and multiple locations in tests.

File: libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts (hunk 12 above)
Comment (griest024) [reply libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts]: The errors are shared by all the different driver actions so we might be clearing errors that are still relevant.

File: libs/cart/src/selectors/cart.selector.ts (hunk 9 above)
Comment (griest024) [reply libs/cart/src/selectors/cart.selector.ts]: Changed in [`58bd413`](https://github.com/graycoreio/daffodil/pull/681/commits/58bd4135400209469b1e66d1f36c7be4f7da7c88).

File: libs/cart/src/effects/cart.effects.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/effects/cart.effects.ts]: Changed in [`0995209`](https://github.com/graycoreio/daffodil/pull/681/commits/0995209afeff022414d8e7f626d3ac39ca677273) and [`14f2376`](https://github.com/graycoreio/daffodil/pull/681/commits/14f23768313d28d145234b39d6c5db3856817254).

File: libs/cart/src/effects/cart.effects.ts [hunk 14]
```diff
@@ -11,64 +11,91 @@ import {
...
+          return new DaffCartCreateSuccess(resp);
+        }),
+        catchError(error => {
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/effects/cart.effects.ts]: Changed in [`0995209`](https://github.com/graycoreio/daffodil/pull/681/commits/0995209afeff022414d8e7f626d3ac39ca677273).

File: libs/cart/src/effects/cart-payment-methods.effects.ts (hunk 13 above)
Comment (griest024) [reply libs/cart/src/effects/cart-payment-methods.effects.ts]: Changed in [`14f2376`](https://github.com/graycoreio/daffodil/pull/681/commits/14f23768313d28d145234b39d6c5db3856817254).

File: libs/cart/src/reducers/cart.reducer.ts [hunk 15]
```diff
@@ -1,49 +1,46 @@
...
-        cart: {
-          ...state.cart,
-          items: [],
-          ...action.payload
-        }
-      };
-    default:
-      return state;
-  }
+export function reducer(
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/reducers/cart.reducer.ts]: Changed in [`5bf412b`](https://github.com/graycoreio/daffodil/pull/681/commits/5bf412b4ff72496bd564601b17b109cd3fd6d62f).

File: libs/cart/src/reducers/cart-state.interface.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import { DaffCart } from '../models/cart';
+
+export interface CartState {
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/reducers/cart-state.interface.ts]: Changed in [`5bf412b`](https://github.com/graycoreio/daffodil/pull/681/commits/5bf412b4ff72496bd564601b17b109cd3fd6d62f).

File: libs/cart/src/reducers/cart-reducers-state.interface.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+import { CartState } from './cart-state.interface';
+
+export interface State {
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/reducers/cart-reducers-state.interface.ts]: Changed in [`5bf412b`](https://github.com/graycoreio/daffodil/pull/681/commits/5bf412b4ff72496bd564601b17b109cd3fd6d62f).

File: libs/cart/src/reducers/cart-reducers.ts [hunk 18]
```diff
@@ -0,0 +1,8 @@
+import { ActionReducerMap } from '@ngrx/store';
+
+import { reducer } from './cart.reducer';
+import { State } from './cart-reducers-state.interface';
+
+export const reducers: ActionReducerMap<State> = {
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/reducers/cart-reducers.ts]: Changed in [`5bf412b`](https://github.com/graycoreio/daffodil/pull/681/commits/5bf412b4ff72496bd564601b17b109cd3fd6d62f).

File: libs/cart/src/effects/cart-shipping-methods.effects.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
+import { Injectable, Inject } from '@angular/core';
+import { switchMap, map, catchError, mapTo, tap } from 'rxjs/operators';
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/effects/cart-shipping-methods.effects.ts]: Changed in [`14f2376`](https://github.com/graycoreio/daffodil/pull/681/commits/14f23768313d28d145234b39d6c5db3856817254).

File: libs/cart/src/effects/cart.effects.spec.ts (hunk 8 above)
Comment (griest024) [reply libs/cart/src/effects/cart.effects.spec.ts]: Changed in [`2eea542`](https://github.com/graycoreio/daffodil/pull/681/commits/2eea54244a04f5331df801d853e83d2dd2c74ab2).

File: libs/cart/src/effects/cart.effects.spec.ts (hunk 7 above)
Comment (griest024) [reply libs/cart/src/effects/cart.effects.spec.ts]: Changed in [`2eea542`](https://github.com/graycoreio/daffodil/pull/681/commits/2eea54244a04f5331df801d853e83d2dd2c74ab2).

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts (hunk 5 above)
Comment (griest024) [reply libs/cart/src/reducers/cart/cart.reducer.spec.ts]: Changed in [`2eea542`](https://github.com/graycoreio/daffodil/pull/681/commits/2eea54244a04f5331df801d853e83d2dd2c74ab2).

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts (hunk 3 above)
Comment (griest024) [reply libs/cart/src/reducers/cart/cart.reducer.spec.ts]: Changed in [`2eea542`](https://github.com/graycoreio/daffodil/pull/681/commits/2eea54244a04f5331df801d853e83d2dd2c74ab2).

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts (hunk 1 above)
Comment (griest024) [reply libs/cart/src/reducers/cart/cart.reducer.spec.ts]: Changed in [`2eea542`](https://github.com/graycoreio/daffodil/pull/681/commits/2eea54244a04f5331df801d853e83d2dd2c74ab2).

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts (hunk 4 above)
Comment (griest024) [reply libs/cart/src/reducers/cart/cart.reducer.spec.ts]: Changed in [`435f5af`](https://github.com/graycoreio/daffodil/pull/681/commits/435f5af42647263c886b9f9c6edbd0d52f35af15).

File: libs/cart/src/reducers/cart/cart.reducer.spec.ts (hunk 2 above)
Comment (griest024) [reply libs/cart/src/reducers/cart/cart.reducer.spec.ts]: Changed in [`4dcc86c`](https://github.com/graycoreio/daffodil/pull/681/commits/4dcc86c5406fda6bab2ad536c6f3bdbf8e4d2111).

File: libs/cart/src/actions/cart-shipping-information.actions.ts [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+export class DaffCartShippingInformationUpdateFailure implements Action {
+  readonly type = DaffCartShippingInformationActionTypes.CartShippingInformationUpdateFailureAction;
+
+  constructor(public payload: string) {}
+}
+
+export class DaffCartShippingInformationDelete implements Action {
+  readonly type = DaffCartShippingInformationActionTypes.CartShippingInformationDeleteAction;
+
+  constructor(public id?: string | number) {}
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/actions/cart-shipping-information.actions.ts]: Changed in [`4300da6`](https://github.com/graycoreio/daffodil/pull/681/commits/4300da64c7e4bde91010a1a3c220520509c6a70b).

**Comment by griest024** [comment]
@lderrickable I added a `fromCart` and `fromCartReducer` export in [`a0c118a`](https://github.com/graycoreio/daffodil/pull/681/commits/a0c118a654f8cda38eda02ea396983097e6a8576) to try and reduce the breaking changes. LMK what you think.

File: libs/cart/src/reducers/cart-shipping-information/cart-shipping-information.reducer.ts (hunk 11 above)
Comment (lderrickable) [reply libs/cart/src/reducers/cart-shipping-information/cart-shipping-information.reducer.ts]: Ok. I don't understand fully what's going on here but I added it to the list of things to take care of later.

File: libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts (hunk 12 above)
Comment (lderrickable) [reply libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts]: Good point. I think we need a more nuanced errors object then, because if we don't clear the errors on success, we won't know when there's really an error or not. I think a good solution could something like: 
```
export interface DaffCartReducerState {
  cart: DaffCart,
  loading: boolean,
  errors: DaffCartError[]
}

export interface DaffCartError {
   type: DaffCartErrorTypeEnum;
   errors: string[];
}

export enum DaffCartErrorTypeEnum {
   ShippingAddress = "Shipping Address",
   Cart = "Cart",
   CartItem = "Cart Item",
   ShippingMethod = "Shipping Method",
   ...
}
```
then in the reducers do:
```
export function reducer(
  state = initialState,
  action: ActionTypes
): DaffCartReducerState {
  switch (action.type) {
    case DaffCartActionTypes.CartCreateSuccessAction:
      return {
        ...state,
        cart: {
          ...state.cart,
          ...action.payload
        },
        loading: false,
        errors: state.errors.filter(error => error.type !== DaffCartErrorTypeEnum.Cart)
      };

    case DaffCartActionTypes.CartCreateFailureAction:
      return {
        ...state,
        loading: false,
        errors: state.errors.concat({ type: DaffCartErrorTypeEnum.Cart, error: action.payload })
      };

    default:
      return state;
  }
}
```
Then people could filter on different feature areas and get specific errors. We could also make selectors for getting errors for specific feature areas. That'd be cool.

File: libs/cart/src/facades/cart/cart.facade.ts [hunk 21]
```diff
@@ -4,19 +4,61 @@ import { Action, Store, select } from '@ngrx/store';
...
 export class DaffCartFacade implements DaffStoreFacade<Action> {
-
   loading$: Observable<boolean>;
   cart$: Observable<DaffCart>;
-  
-  constructor(private store: Store<fromCart.State>) {
-    this.loading$ = this.store.pipe(select(fromCart.selectCartLoadingState));
-    this.cart$ = this.store.pipe(select(fromCart.selectCartValueState));
+  errors$: Observable<string[]>;
+  cartId$: Observable<DaffCart['id']>;
```
Comment (lderrickable) [inline libs/cart/src/facades/cart/cart.facade.ts]: you can remove all of the "cart" prefixes. Since this is coming from the cart facade, it should be assumed that all these properties pertain to cart.

File: libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts (hunk 12 above)
Comment (griest024) [reply libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts]: What about storing the errors in an object keyed by the enum? i.e.
```ts
export interface DaffCartReducerState {
  cart: DaffCart,
  loading: boolean,
  errors: DaffCartErrors
}

export interface DaffCartErrors {
   [key: DaffCartErrorTypeEnum]: string[]
}
```

It would make selecting for types of errors easier and more performant.

File: libs/cart/src/facades/cart/cart.facade.ts (hunk 21 above)
Comment (griest024) [reply libs/cart/src/facades/cart/cart.facade.ts]: Changed in [`fdf6e8b`](https://github.com/graycoreio/daffodil/pull/681/commits/fdf6e8b2be945c8ad297e46ca13638b50cfc2191).

File: libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts (hunk 12 above)
Comment (lderrickable) [reply libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts]: sure, that seems better

Comment (griest024) [reply libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts]: Changed in [`42f64c5`](https://github.com/graycoreio/daffodil/pull/681/commits/42f64c5c33b8d2ce7071d15a299d7cab2ef0e23f).

Comment (griest024) [reply libs/cart/src/reducers/cart-shipping-address/cart-shipping-address.reducer.ts]: Changed to do error reducing logic in separate methods in [`d8b7537`](https://github.com/graycoreio/daffodil/pull/681/commits/d8b75371b3f15b19a655a63800913bb975dbb614).

File: libs/cart/src/reducers/cart-billing-address/cart-billing-address.reducer.ts [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+): DaffCartReducerState {
+  switch (action.type) {
+    case DaffCartBillingAddressActionTypes.CartBillingAddressLoadAction:
+    case DaffCartBillingAddressActionTypes.CartBillingAddressUpdateAction:
+      return { ...state, loading: true };
+
+    case DaffCartBillingAddressActionTypes.CartBillingAddressLoadSuccessAction:
+      return {
+        ...state,
+        ...resetErrors(state),
```
Comment (lderrickable) [inline libs/cart/src/reducers/cart-billing-address/cart-billing-address.reducer.ts]: Either a lot of these `resetErrors` need tested or all of them do.

File: libs/cart/src/facades/cart/cart.facade.ts [hunk 23]
```diff
@@ -4,19 +4,88 @@ import { Action, Store, select } from '@ngrx/store';
...
 export class DaffCartFacade implements DaffStoreFacade<Action> {
-
   loading$: Observable<boolean>;
   cart$: Observable<DaffCart>;
-  
-  constructor(private store: Store<fromCart.State>) {
-    this.loading$ = this.store.pipe(select(fromCart.selectCartLoadingState));
-    this.cart$ = this.store.pipe(select(fromCart.selectCartValueState));
+
+  cartErrors$: Observable<DaffCartErrors[DaffCartErrorType.Cart]>;
```
Comment (lderrickable) [inline libs/cart/src/facades/cart/cart.facade.ts]: I think we should keep the overall errors property too

**Comment by lderrickable** [comment]
ok that should be all. I'll just look at the new commits you make and we should be good.

File: libs/cart/src/reducers/cart-billing-address/cart-billing-address.reducer.ts (hunk 22 above)
Comment (griest024) [reply libs/cart/src/reducers/cart-billing-address/cart-billing-address.reducer.ts]: They're not being exported and their logic is tested with the reducer spec.

Comment (lderrickable) [reply libs/cart/src/reducers/cart-billing-address/cart-billing-address.reducer.ts]: I don't see this being tested in the reducer spec:
```
  describe('when CartShippingAddressLoadSuccessAction is triggered', () => {
    let result;
    let state: DaffCartReducerState;

    beforeEach(() => {
      state = {
        ...initialState,
        loading: true
      }

      const cartListLoadSuccess = new DaffCartShippingAddressLoadSuccess(cart.shipping_address);

      result = reducer(state, cartListLoadSuccess);
    });

    it('should indicate that the cart is not loading', () => {
      expect(result.loading).toEqual(false);
    });

    it('should set shipping_address from action.payload', () => {
      expect(result.cart.shipping_address).toEqual(cart.shipping_address)
    });
  });
```

Comment (griest024) [reply libs/cart/src/reducers/cart-billing-address/cart-billing-address.reducer.ts]: Oh you're right, my bad. I forgot to add tests covering the error reset in the first place.

File: libs/cart/src/facades/cart/cart.facade.ts (hunk 23 above)
Comment (griest024) [reply libs/cart/src/facades/cart/cart.facade.ts]: Changed in [`6128980`](https://github.com/graycoreio/daffodil/pull/681/commits/612898086944308222bcf7f16e9361fb4c2b4ebf).

File: libs/cart/src/effects/cart-shipping-information.effects.ts [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+        map((resp: V) => new DaffCartShippingInformationUpdateSuccess(resp)),
+        catchError(error => of(new DaffCartShippingInformationUpdateFailure('Failed to update cart shipping information')))
+      )
+    )
+  )
+
+  @Effect()
+  delete$ = this.actions$.pipe(
+    ofType(DaffCartShippingInformationActionTypes.CartShippingInformationDeleteAction),
+    switchMap((action: DaffCartShippingInformationDelete<DaffCartShippingRate>) =>
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/effects/cart-shipping-information.effects.ts]: Changed in [`47ef8ca`](https://github.com/graycoreio/daffodil/pull/681/commits/47ef8ca051443e15c3a15b8723248e583b3e497c).

File: libs/cart/src/reducers/cart-billing-address/cart-billing-address.reducer.ts (hunk 22 above)
Comment (griest024) [reply libs/cart/src/reducers/cart-billing-address/cart-billing-address.reducer.ts]: Changed in [`55e630e`](https://github.com/graycoreio/daffodil/pull/681/commits/55e630edaee54d8d47ecdd97955efe1c63ecefdd).

File: libs/cart/src/reducers/cart.reducer.ts [hunk 25]
```diff
@@ -1,49 +1,46 @@
-import { DaffCartActionTypes, DaffCartActions } from '../actions/cart.actions';
-import { DaffCart } from '../models/cart';
+import { reducer as cartReducer } from './cart/cart.reducer';
+import { reducer as cartItemReducer } from './cart-item/cart-item.reducer';
+import { reducer as cartBillingAddressReducer } from './cart-billing-address/cart-billing-address.reducer';
+import { reducer as cartShippingAddressReducer } from './cart-shipping-address/cart-shipping-address.reducer';
+import { reducer as cartShippingMethodsReducer } from './cart-shipping-methods/cart-shipping-methods.reducer';
+import { reducer as cartShippingInformationReducer } from './cart-shipping-information/cart-shipping-information.reducer';
+import { reducer as cartPaymentReducer } from './cart-payment/cart-payment.reducer';
+import { reducer as cartPaymentMethodsReducer } from './cart-payment-methods/cart-payment-methods.reducer';
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/src/reducers/cart.reducer.ts]: Changed in [`55e630e`](https://github.com/graycoreio/daffodil/pull/681/commits/55e630edaee54d8d47ecdd97955efe1c63ecefdd).

**Comment by lderrickable** [comment]
lgtm
